File: assignment_p1.py
__author__ = 'urajkuma@ucsd.edu,A91060509,yil261@ucsd.edu,A91085115,L1Kong@ucsd.edu,A97010449'

from Queue import Queue

million = 1000000
prime = [False]*million

def sieve():
    for i in range(2, 1000, 1):
        if prime[i] == False:
            for j in range(i*2, million, i):
                prime[j] = True
    
def numDigits(num):
    return len(str(num))

def arrToNum(digits, size):
    temp = 0
    x = 0
    while x < size:
        temp = temp*10 + digits[x]
        x+=1

    return temp

def arr(digits, num):
    x = numDigits(num) - 1
    while num:
        digits[x] = num%10
        num = num/10
        x -= 1

def exponents(size):
    threshold = 1
    while size > 1:
        threshold *= 10
        size -= 1

    return threshold




def main():
    b,a = map(int, raw_input().split())
    size = numDigits(a)
    digits = [None]*size
    primes = Queue()
    primes.put(a)
    threshold = exponents(size)
    sieve()
    parent = [-1]*million
    distance = [-1]*million
    parent[a] = 0;
    distance[a] = 0    
    while (primes.empty() != True):
        num = primes.get()
        for x in range((size-1), -1, -1):
            arr(digits, num)
            for i in range(0, 10):
                digits[x] = i
                temp = arrToNum(digits, size)
                if ((prime[temp] == False) and (distance[temp] == -1) and (temp >= threshold)):
                    distance[temp] = distance[num] + 1
                    parent[temp] = num
                    primes.put(temp)

    if(distance[b] == -1):
        print "UNSOLVABLE"
    else:
        while(parent[b] != False):
            print b,
            b = parent[b]
        print a


if __name__ == '__main__':
    main()

File: assignment_p1.cpp
//implementation of the assignment in c++ for reference
#include <bits/stdc++.h>
using namespace std;
static int million = 100000;
bool prime[100000]; //list of primes up to 1,000,000

//create list of primes up to 1,000,000
void sieve()
{
	//indices that are primes will stay as false
    for(int i=2; i*i< million; i++)
    {
        if(!prime[i])
        {
            for(int j=i*2; j< million; j += i)
                prime[j] = true; 
        }
    }
}

//number of digits in a number
int numDigits(int number)
{
    int digits = 0;
    while (number) {
        number /= 10;
        digits++;
    }
    return digits;
}

//conver array back to number
int arrToNum(int digits[], int size)
{
    int temp = 0, x = 0;
    while(x < size)
        temp = temp*10 + digits[x++];

    return temp;
}

void arr(int digits[], int num)
{
    int x = numDigits(num)-1;
    while(num)
    {
        digits[x--] = num%10;
        num/=10;
    }
}

int exponents(int size)
{
	int threshold = 1;

	while(size-- > 1)
		threshold *= 10;

	return threshold;
}

int main()
{
	sieve(); //create the list of primes
	int a, b, parent[100000], distance[100000];
	scanf("%d %d", &a, &b);
	int size = numDigits(a);
	int  digits[size];
	memset(distance, -1, sizeof(distance));
    memset(parent, -1, sizeof(parent));
    queue<int> primes;
    distance[a]=0; 
    primes.push(a); //queue for bfs
    int threshold = exponents(size);//to ensure additions and deletions don't occue
    parent[a]=0;
    
    while(!primes.empty()){

    	int num = primes.front();
    	primes.pop();
    	for(int x = (size-1); x >= 0; x--){

    		arr(digits, num);
    		for(int i = 0; i <= 9; i++){

    			digits[x] = i;
    			int temp = arrToNum(digits, size);
    			//last condition is to ensure additions/removals of numbers don't occur
    			if((!prime[temp]) && distance[temp]==-1 && temp >= threshold) 
    			{
    				distance[temp] = distance[num] + 1;
    				parent[temp] = num;
    				primes.push(temp);
    			}
    		}
    	}
    }
    if(distance[b]==-1) 
        cout << "UNSOLVABLE" << endl; 
    else
    {
        while(parent[b] != 0)
        {
            cout << b << " ";
            b = parent[b];
        }
    }
    return 0;

}
